File: Java/TwoSum.java
//------------------------------PROBLEM 1--------------------------------//
//                               TWO SUM                                 //


//Logic: 
// -> Create a HashMap to store the numbers and their indices
// -> Iterate through the array of numbers
// -> For each number, calculate its complement (target - current number)
// -> If the complement exists in the HashMap, return the indices
// -> If not, add the current number and its index to the HashMap
// -> If no solution is found, return an empty array


import java.util.HashMap;

class Solution {

    public int[] twoSum(int[] nums, int target) {

        HashMap<Integer, Integer> map = new HashMap<>(); 
        
        for (int i = 0; i < nums.length; i++) {

            int complement = target - nums[i];
            
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            
            map.put(nums[i], i);
        }
        
        return new int[] {};
    }
}


//Time Complexity:
// -> Iterating through the array: O(n)
// -> HashMap operations (put and get): O(1)
// Overall, O(1) * n
// => O(n)

File: Python/TwoSum.py
#------------------------------PROBLEM 1--------------------------------#
#                               TWO SUM                                 #


# Logic:
# -> Loop through each element in the list
# -> For each element, loop through the rest of the list to find a pair that sums to the target
# -> If such a pair is found, return their indices
# -> If no such pair exists, return an empty list


class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        
        n = len(nums)

        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] + nums[j] == target:
                    return [i, j]
        
        return []


# Time Complexity:
# -> Outer loop runs n times
# -> Inner loop runs (n - i - 1) times for each i
# [Total comparisons: n + (n-1) + (n-2) + ... + 1 = n(n-1)/2]
# Overall, O(n) * O(n)
# => O(n^2)
